src/Compilers/VisualBasic/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncMethodToClassRewriter.Await.vb

```diff
@@ -142,86 +142,89 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                            asNotifyCompletion)
-
-                    ' >>>>   builder.AwaitOnCompleted(Of TAwaiter,TSM)((ByRef) dcast2, (ByRef) Me)
-                    Dim awaitOnCompletedCall As BoundStatement =
-                        Me.F.ExpressionStatement(
-                            Me._owner.GenerateMethodCall(
-                                builderFieldAsRValue,
-                                Me._owner._builderType,
-                                "AwaitOnCompleted",
-                                ImmutableArray.Create(Of TypeSymbol)(asNotifyCompletion.Type, Me.F.Me().Type),
-                                {notifyCompletionAsLValue, Me.F.ReferenceOrByrefMe()}))
-
-                    blockBuilder.Add(
-                        Me.F.Block(
-                            ImmutableArray.Create(Of LocalSymbol)(asCriticalNotifyCompletion, asNotifyCompletion),
-                            asCriticalNotifyCompletionAssignment,
-                            Me.F.If(
-                                condition:=Me.F.Not(Me.F.ReferenceIsNothing(Me.F.Local(asCriticalNotifyCompletion, False))),
-                                thenClause:=awaitUnsafeOnCompletedCall,
-                                elseClause:=Me.F.Block(
-                                    asNotifyCompletionAssignment,
-                                    awaitOnCompletedCall))))
-
-                Else
-                    ' STMT:   this.builder.AwaitUnsafeOnCompleted(Of TAwaiter,TSM)((ByRef) $awaiterTemp, (ByRef) Me)
-                    '  or
-                    ' STMT:   this.builder.AwaitOnCompleted(Of TAwaiter,TSM)((ByRef) $awaiterTemp, (ByRef) Me)
-                    Dim useUnsafeOnCompleted As Boolean =
-                        Conversions.IsWideningConversion(
-                            Conversions.ClassifyDirectCastConversion(
-                                awaiterType,
-                                Me.F.Compilation.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_ICriticalNotifyCompletion),
-                                useSiteDiagnostics:=Nothing))
-
-                    blockBuilder.Add(
-                        Me.F.ExpressionStatement(
-                            Me._owner.GenerateMethodCall(
-                                builderFieldAsRValue,
-                                Me._owner._builderType,
-                                If(useUnsafeOnCompleted, "AwaitUnsafeOnCompleted", "AwaitOnCompleted"),
-                                ImmutableArray.Create(Of TypeSymbol)(awaiterType, Me.F.Me().Type),
-                                {Me.F.Local(awaiterTemp, True), Me.F.ReferenceOrByrefMe()})))
+                Dim ICriticalNotifyCompletion = Me.F.WellKnownType(WellKnownType.System_Runtime_CompilerServices_ICriticalNotifyCompletion)
+
+                If Not ICriticalNotifyCompletion.IsErrorType Then
+                    If isAwaiterTempImpliesLateBound Then
```

Comment:
Aside from wrapping the code with an extra condition, the only non-space changes in the body are on line 156 and 216.